### ISSUE TITLE
Android port?

### ISSUE THREAD

**Issue by merpysoup** [OpenXRay/xray-16#643]
After project s.t.a.l.k.e.r died this is the biggest hope for an android port, which would be awesome. Morrowind has had an Android port for a very long time, so I believe that an android port will create plenty of publicity and allow more people to start working on OpenXRay. 

Android runs on opengl and vulkan mostly, right? and plenty of Linux operating systems are ARM based so I think an Android port would be very awesome.

Android port will create more publicity and supporters for the project, and will give windows users a reason to check this out.

**Comment by ghost** [comment]
Morrowind exists as open source projects with many coders who help to implement stuff. But thinking about that, they are re implement all stuff just from scratch. For the stalker this experience doesn't work, because OpenXRay team need more programmers or something else. 

Just do port on this engine can't be real at this point. Someday and someone...

**Comment by lonelyfootpad** [comment]
I've tested Morrowind on android but I can't say I've played it. I doubt I'd play something like this on android either. I don't know if it is even feasible to port something like this. It'd probably have to be a separate project that is a dedicated build from scratch to recreate the game for android rather than trying to port this codebase.

**Comment by kotleni** [comment]
**I'm working on it.** I managed to run the engine to the main menu.

**Comment by kotleni** [comment]
> После смерти проекта stalker это самая большая надежда на порт для Android, и это было бы здорово. В Morrowind уже очень давно есть порт для Android, поэтому я считаю, что порт для Android создаст много рекламы и позволит большему количеству людей начать работать над OpenXRay.
> 
> Android в основном работает на opengl и vulkan, верно? и многие операционные системы Linux основаны на ARM, поэтому я думаю, что порт Android был бы очень крутым.
> 
> Перенос Android повысит популярность проекта и поддержит его, а также даст пользователям Windows повод проверить это.

**That's right, Android works with OpenGL - but its mobile version (OpenGLES). Moving from OpenGL to OpenGLES rendering is not an easy task.
If you make it work on a smartphone, then only write your own implementation.**

Nobody wants to do this for free.

**Comment by Xottab-DUTY** [comment]
I'm gradually working on refactoring renderers for unifying, modernizing (e.g. getting rid of D3DX) and improvement (e.g. proper frame graph instead of `CRenderTarget` class) purposes. Someday, I'll get to OpenGL ES as a consequence of these refactorings...

By the way, a man from our discord managed to run the game on ARM tablet. That tablet had Linux installed, not Android, but it's still a huge progress!

**Comment by kotleni** [comment]
> Я постепенно работаю над рефакторингом рендереров для унификации, модернизации (например, избавления от D3DX) и улучшения (например, правильного графа кадра вместо `CRenderTarget`класса) целей. Когда-нибудь я доберусь до OpenGL ES в результате этих рефакторингов ...
> 
> Кстати, человеку из нашего раздора удалось запустить игру на планшете ARM. На этом планшете был установлен Linux, а не Android, но это все еще огромный прогресс!

_Насчет arm сборки, можешь связать меня с этим человеком?_

**Comment by Xottab-DUTY** [comment]
> By the way, a man from our discord managed to run the game on ARM tablet. That tablet had Linux installed, not Android, but it's still a huge progress!

> Насчет arm сборки, можешь связать меня с этим человеком?

Screenshot by @KiralyCraft, he is the person who managed to run the engine :)

![OpenXRay on ARM Linux](https://media.discordapp.net/attachments/489453264157409281/831473803724587028/20210413_131806.png?width=888&height=666)
*[(link to the message)](https://discord.com/channels/410170555619082240/489453264157409281/831473811936641024)*

**Comment by davie504** [comment]
Неплохо,получается просто прилинковать транслятор и порт готов

**Comment by GerKaYZeR** [comment]
Доброго времени суток Товарищи! Проект жив? Я просто хотел бы примкнуть к вашей команде в программировании я конечно слаб но в дизайне что то и могу просто стоит показать что надо делать и я справлюсь. Кто то говорил что нужно конвертировать с OpenGL в OpenGLES так я могу бесплатно свободного времени у меня каждый вечер.

**Comment by Godlux** [comment]
https://www.youtube.com/watch?v=HiikRqjkhdE

Can we do something with this to make this port happen?

**Comment by Xottab-DUTY** [comment]
> [youtube.com/watch?v=HiikRqjkhdE](https://www.youtube.com/watch?v=HiikRqjkhdE)
> 
> Can we do something with this to make this port happen?

With this – nothing. The video just shows the original game being run in some kind of emulator. (technically it's not, but for simplicity I'll call it an emulator)
To make it "not emulated", we need to port the code to Android and implement proper touch screen support and proper input controls. That's a bunch of programming work.

**Comment by Godlux** [comment]
Yeah, part of "not emulator" means that maybe we can use this like a docker (environment to run OpenXRay). There is a arm wine, OpenXRay can be built for arm64 (i've done this for my NS). So the question is "can we using this tools create environment to run OpenXRay as native as possible, avoiding bunch of programming work?"

**Comment by Xottab-DUTY** [comment]
> Yeah, part of "not emulator" means that maybe we can use this like a docker (environment to run OpenXRay). There is a arm wine, OpenXRay can be built for arm64 (i've done this for my NS). So the question is "can we using this tools create environment to run OpenXRay as native as possible, avoiding bunch of programming work?"

Weell! That's interesting!

I think this should be possible then. But I have no experience with Docker and Android overall.
Maybe there's someone lurking around for an interesting challenge to do... :D

**Comment by strikersix23** [comment]
I was just reading through this and saw the opengl part for that there's gl4es which as an wrapper library.
Source code is here 

https://github.com/ptitSeb/gl4es

It's just a suggestion to use it as an alternative to fully porting the renderer to opengl

**Comment by radiKal07** [comment]
The engine uses DirectX, wouldn't it be easier to use something like DXVK to target Vulkan? Vulkan already works on Android. Wouldn't this be easier than re-writing the engine to use OpenGL ES instead of OpenGL?

**Comment by kotleni** [comment]
> The engine uses DirectX, wouldn't it be easier to use something like DXVK to target Vulkan? Vulkan already works on Android. Wouldn't this be easier than re-writing the engine to use OpenGL ES instead of OpenGL?

You're right! It's true, and useful not only for Android port :)

**Comment by strikersix23** [comment]
> The engine uses DirectX, wouldn't it be easier to use something like DXVK to target Vulkan? Vulkan already works on Android. Wouldn't this be easier than re-writing the engine to use OpenGL ES instead of OpenGL?

The issue is it's going to be converting an older version not to mention dxvk is an native adaptation to windows/pe librarys and can use  dolls to link to wine. The issues with dxvk would be an an rewrite to the dxvk itself even with with the steam sniper build(native library and headers) but you would most likely be running on an unsupported Vulcan version dxvk requires 1.3 while most androids use 1.1 unless would have an adreno gpu which thanks to adreno tools would be possible to run 1.3 Vulkan without su. Which means using a dynamic driver interface for android

It would take a lot either way

**Comment by strikersix23** [comment]
> > The engine uses DirectX, wouldn't it be easier to use something like DXVK to target Vulkan? Vulkan already works on Android. Wouldn't this be easier than re-writing the engine to use OpenGL ES instead of OpenGL?
> 
> You're right! It's true, and useful not only for Android port :)

Basic level yes. Actually implementation no

**Comment by sandstranger** [comment]
Портированный движок под Android с regal находится тут.
https://github.com/sandstranger/xray-16-android-port
https://github.com/gabrielcuvillier/regal-emscripten-fixes
Сейчас падает на компиляции шейдеров, что в общем-то ожидаемо
Все шейдеры надо переделывать под gles2

**Comment by kotleni** [comment]
@sandstranger Выглядит не плохо, но есть реально смысл портировать его через **GLES**? Вроде как уже обсуждали что сперва **Vulkan** нужно затащить в проект >.<?

**Comment by sandstranger** [comment]
> @sandstranger Выглядит не плохо, но есть реально смысл портировать его через **GLES**? Вроде как уже обсуждали что сперва **Vulkan** нужно затащить в проект >.<?

Нет смысла и нереально перепиливывать все шейдеры
За зинком с вулканом под ведро сюда
https://github.com/SolDev69/unified-mesa-project
https://github.com/SolDev69/SolCraftLauncher

**Comment by klaxons1** [comment]
Согласен, вулкан на вёдрах чаще всего текстуры dxt умеет аппаратно декодить, в отличии от глес

**Comment by KiralyCraft** [comment]
I've managed to get the latest git commit working on my Galaxy S9 with LineageOS 20, but the performance is really bad. I'm running in an Arch Linux chroot container with Termux:X11 as the display client and no desktop environment. Hardware acceleration is working, and the average FPS seems to be around 2. In the menu, I get around 8 FPS. 

All video settings are set to the lowest possible value, no AA or anything else. The render resolution is 1184x576.

There is still plenty of free RAM for the game, so I assume it's just a graphics thing. 

The "emulator" above seems to have better performance, but this one is running raw OpenGL
![URTBZeXlGdCsnfptyDZThaqd](https://github.com/user-attachments/assets/84584e35-12dd-4439-8507-54f373d5c8a9)

**Comment by kotleni** [comment]
@KiralyCraft I think it's uses some kind of opengl to gles converter is minimum? And additionally native port will be better anyway.

**Comment by strikersix23** [comment]
GL4ES is pretty fast its in Open Morrowind Mobile port currently maintened by sisha2 on github. It's also created by the Box64/Box86 dev to work on arm64 including android

**Comment by strikersix23** [comment]
> Портированный движок под Android с regal находится тут. https://github.com/sandstranger/xray-16-android-port https://github.com/gabrielcuvillier/regal-emscripten-fixes Сейчас падает на компиляции шейдеров, что в общем-то ожидаемо Все шейдеры надо переделывать под gles2

You could use ANGLE or use SPIRV (something similer I forgot the exact program) but it would help. Angle could help in debugging the shaders if I'm not mistaken

**Comment by sandstranger** [comment]
дошел до рендеринга через зинк, рендерится все через SDLSURFACE, но сейчас рендеринг работает крайне нестабильно, выдает минимум фпс и падает
чатгпт рекоменует SDLRENDER.
Пример рендера с моего планшета на ведре
https://imgur.com/a/O2kOzdG

**Comment by kotleni** [comment]
@sandstranger Good, btw game is running without crashes? Or just only main menu?

**Comment by sandstranger** [comment]
> @sandstranger Good, btw game is running without crashes? Or just only main menu?

only main menu lol with fps about zero, because I am using SDLSURface for rendering.
I need to rewrite this part
